### ISSUE TITLE
Add setuptools_scm support

### ISSUE THREAD

**Issue by Czaki** [beeware/briefcase#474]
In my project I use setuptools_scm for versioning (it calculate version from git tag and create proper file handling it). My project is distributed as python package and pyinstaler executable. I would like to switch from pyinstaller to briefcase (possibility of pip support) but I do not want to manually change version in file. 

**Describe the solution you'd like**
Add support to read version from setuptools_scm instead of manual update

**Describe alternatives you've considered**
Manual updating version. read version from file

**Comment by freakboy3742** [comment]
Thanks for the suggestion. I agree that "version from SCM" is a use case worth considering.

There are a couple of complications to what you describe, though. Some packaging schemes have *very* strict version number requirements. Windows only allows strict numerical Triples. Android requires a build "sequence" number. IIRC, iOS only allows numbers. As a result, building an installer "for git hash xxxxx" isn't actually something you're going to be able to do in practice; even "for git tag xxxxx" will have some very serious constraints. We'll need to work out how to manage those constraints in any SCM-based version number solution.

There are also alternatives to `setuptools_scm` out there ([versioneer](https://github.com/warner/python-versioneer) being one that I am aware of); we'll need to make a decision whether to make an opinionated choice of one of those packages, or different version control systems (or even introduce a plugin system to support alternatives).

**Comment by Czaki** [comment]
So maybe add option to provide version by function? 
Something like `get_version.py:get_version` Then user could provide simple wrapper function which provide version in proper format?

**Comment by freakboy3742** [comment]
A user-provided function would be one way to provide compliant version numbers; however, that function would have to either (a) return a PEP508-compliant version number, which is then processed using the existing version number handling code, or (b) return a different version number for each platform.

What we really need at this point is a complete proposal for needs to change, and what the experience will be like as an end-user.

**Comment by Czaki** [comment]
> A user-provided function would be one way to provide compliant version numbers; however, that function would have to either (a) return a PEP508-compliant version number, which is then processed using the existing version number handling code, or (b) return a different version number for each platform.

But is is enough to add description in documentation for this and fail if user function return version in wrong format.

**Comment by mhsmith** [comment]
As mentioned in #1277, there is now a standard `dynamic` setting in pyproject.toml which may be useful for this.

**Comment by andrewleech** [comment]
The official dynamic version support in `pyproject.toml` is handled deep within the setuptools / disutils / Distribution system. I've attempted to trace through how this works to find a simple means of querying it within briefcase, however it doesn't appear to be exposed in any official api as such. 

Essentially:
* `[project]` needs to include `dynamic = ["version"]`
* the `[build-system]` section needs to have the dynamic version providing package listed in its `requires` 
* the dynamic version providing package needs to be enabled with a `[tool.<versioning package>]` block

Once these are done, part of `setuptools.dist.Distributrion` will call through to `setuptools.finalize_distribution_options` registered entry-point in the versioning package which then figures out the real version and sets it on the Distribution object.

However, there's an alternate format of configuring dynamic version without needing a package that supports it, by simply setting a block in pyproject.toml:
```
[tool.setuptools.dynamic]
version = {attr = "__version__.version_py_short"}
```
This one gets flagged with a warning from setuptools that it's still beta functionality. 
If both mechanisms are configured, I'm not sure which one wins.

Basically... if briefcase simply built the app as a wheel or sdist first with the standard tooling (eg. the `build` tool) and then extracted said wheel, then we'd get all the dynamic versioning stuff for free using the official mechanisms.

Trying to replicate the official functionality might be possible, but probably a lot of busywork and I'd also expect it to be somewhat fragile.

**Comment by layday** [comment]
I know the Python packaging landscape is (more than) a little confusing, but there's nothing I'd call "official" dynamic version support in pyproject.toml.   Declaring `version` as dynamic has no significance beyond informing the reader that the version is defined elsewhere.  setuptools does not "own" project configuration, and other build systems might handle extracting the version from VCS or code differently (or not at all) - see e.g. [Flit](https://flit.pypa.io/en/stable/index.html?highlight=__version__#usage), [poetry-dynamic-versioning](https://github.com/mtkennerly/poetry-dynamic-versioning) (based on [dunamai](https://github.com/mtkennerly/dunamai), which is what I'd use if briefcase were to expose a version hook), [hatch-vcs](https://github.com/ofek/hatch-vcs).

**Comment by andrewleech** [comment]
I guess I meant official in the sense of how the current pypa backed tooling uses pyproject.toml to build sdist/wheels.
There's certainly no official version provider package (though setuptools-scm is used as the reference example) but this, my preferred `git-versioner` and all the ones you list that directly support pyproject.toml do so you via "build-system" integration, regardless of whether that build system is setuptools or their own builder.

Briefcase should really be able to work with any dynamic version provider if they all follow a standard pattern... though most of the standard patterns are discounted l designed to ultimately build a wheel.

**Comment by andrewleech** [comment]
That being said, the setuptools workaround pattern could be reused here, providing a section like this:
```
[tool.briefcase.dynamic]
version = {attr = "__version__.version_py_short
```
Which allows the user to define a property/function to call to provide the version.

**Comment by freakboy3742** [comment]
> Briefcase should really be able to work with any dynamic version provider if they all follow a standard pattern... though most of the standard patterns are discounted l designed to ultimately build a wheel.

So - I agree in theory - but I'm not sure there *is* any such thing as a standalone dynamic version provider in the way you're describing it. All the tools I'm aware of exist as plugins to other build systems (setuptools, hatch, poetry etc). It may be possible to adopt one of these tools, but the *configuration* of that tool is going to be something that is Briefcase specific.

I'm also not sure I understand why the ability to configure multiple SCM version number utilities is required. Yes, the ecosystem has a proliferation of near-identical tools... but AFAICT this is mostly due to no single tool gaining traction due to the general mess of the broader Python packaging ecosystem, rather than a demonstration of the highly customised needs of individual versioning schemes.

With that in mind, I make the following suggestion:

If a project only defines `version`, whether that be in the `[project]` PEP621 section, or a `[tool.briefcase.*]` section, it's a hard-coded version number. This is the status quo; we already do validation that the provided string is PEP440 compliant.

If a project defines a PEP621-compliant `dynamic = ["version"]`, `version` in `[tool.briefcase.*]` is allowed to be:

* `{"file": "path/to/file"}`, will grep the contents of the named file, relative to `pyproject.toml`, looking for `__version__ = "(.*)"`, and use the value match (returning an error if the file can't be found, or the regex doesn't match. PEP440 validation is done on this string.
* `{"attr": "myapp.module.__version__"}` will do importlib calls to import "myapp.module" and retrieve the `__version__` attribute, using the same PYTHONPATH used to run `briefcase dev`. PEP440 validation is also done on this string.

If `version` is undefined at the briefcase level, `setuptools_scm` is used to extract the version number, using the path containing `pyproject.toml` as a root. 

The decision to use setuptools_scm is based entirely on the fact that Briefcase itself uses setuptools_scm, and it's the implementation that is used by a lot of the other build tool plugins (e.g., hatch-vcs is really just a wrapper around setuptools_scm). If someone wants to make a particularly impassioned plea to use a different library, I'm open to reasonable discussions, but this would probably need to be paired with Briefcase (and other BeeWare projects) making the same choice.

The decision to make setuptools_scm the default if `version` is undefined is that "discover what you can from the VCS" seems like a reasonable default fallback position, and it requires no additional configuration.

In theory, we *could* expose a plugin interface - allow anyone to define a `briefcase-myversionplugin` tool that registers their own versioning scheme... but I wouldn't consider that necessary for a v1. I'd rather deploy a 95% solution and see if there's demand for the last 5%.

**Comment by layday** [comment]
> I'm also not sure I understand why the ability to configure multiple SCM version number utilities is required. Yes, the ecosystem has a proliferation of near-identical tools... [...]

There are very few SCM version number utilities (I can only think of two: `setuptools_scm` and `dunamai`). What's a preponderence of is build system plug-ins.  There's no unified build system plug-in interface (although there is a [proposal](https://discuss.python.org/t/proposal-for-dynamic-metadata-plugins/25237) for one), so people have got to write separate plug-ins for each build system which supports dynamic metadata. `setuptools_scm` is both an SCM version provider and a setuptools plug-in.  `hatch-vcs` uses `setuptools_scm` under the hood to extract SCM versions and I assume other build systems do too.

> If a project defines a PEP621-compliant `dynamic = ["version"]`, `version` in `[tool.briefcase.*]` is allowed to be: [...]

Just to note, and assuming these are inspired by setuptools, the `file` attribute in setuptools simply loads the contents of a text file. `attr` AST-parses the module to avoid import side-effects (importing other modules).

**Comment by freakboy3742** [comment]
> > I'm also not sure I understand why the ability to configure multiple SCM version number utilities is required. Yes, the ecosystem has a proliferation of near-identical tools... [...]
> 
> There are very few SCM version number utilities (I can only think of two: `setuptools_scm` and `dunamai`). What there is a preponderence of is build system plug-ins. 

That would seem to reduce the impetus for a Briefcase plugin even further.

> Just to note, and assuming these are inspired by setuptools, the `file` attribute in setuptools simply loads the contents of a text file. `attr` AST-parses the module to avoid import side-effects (importing other modules).

"Inspired by setuptools" was the intention, so I agree we should follow those implementations.

I know the `attr` parsing was the subject of an old issue, but wasn't aware it had been resolved. Either way, I agree a parsing-based, side-effect free implementation is preferable.

**Comment by mhsmith** [comment]
Renamed to reflect other possible types and sources of metadata, such as this example from @DaloroAT in #1432:

-----

I have my own project where I'm using [Dynamic Metadata](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata) for the version field. `pyproject.toml` looks like

```yaml
[project]
name = "my-project"
dynamic = ["version"]

[tool.setuptools.dynamic]
version = {file = "my-project/VERSION"}
```

I have the file `VERSION` in the source folder of `my-project` with a version value as a plain string. During package installation, this version was substituted.

**Comment by akaIDIOT** [comment]
I ran into this issue / missing feature while trying to use briefcase to build my project, that defines a dynamic version in `pyproject.toml`, to be resolved by the build backend under normal circumstances. For this, briefcase is a tool in the chain, not my build backend for the entire project, and as such, briefcase not 'following' this approach is a hassle. 

While I get that briefcase would require some more metadata than other tools, would it make sense to invoke a project's build backend as a fallback when some values are declared to dynamic or briefcase detects that it's missing some metadata to function? I'm thinking of [`backend.prepare_metadata_for_build_wheel`](https://peps.python.org/pep-0517/#prepare-metadata-for-build-wheel) in particular. That function is optional according to the spec, but would generate a version in my case, and likely anything declared to be dynamic in the general case. If the declared backend does not support that particular function, there's always the current approach of raising an error.

**Comment by akaIDIOT** [comment]
Fiddling with my own idea mentioned above, it seems the patch below seems to work as a poor man's version of this, using `build` (already declared as a dependency for briefcase) to generate a project's metadata and writing values from that metadata into the project's metadata before calling `merge_pep621_config` with that updated dict. 

The translation from the header values in the `METADATA` file (RFC822-like if I'm not mistaken) is done using `key.title()`, which is likely way shortsighted, but works for my particular use case ("Version" → "version"). As such, this is not PR-ready, but might be useful to someone or to briefcase. I'd be willing to put in the effort to turn this into something merge-worthy, provided it's an avenue worth pursuing at all, please advise 🙏🏻 

```patch
diff --git a/src/briefcase/config.py b/src/briefcase/config.py
index 8d5cc75d..20be61f2 100644
--- a/src/briefcase/config.py
+++ b/src/briefcase/config.py
@@ -8,6 +8,7 @@ import unicodedata
 from pathlib import Path
 from urllib.parse import urlparse
 
+from build.util import project_wheel_metadata
 from packaging.licenses import InvalidLicenseExpression, canonicalize_license_expression
 from packaging.version import InvalidVersion, Version
 
@@ -1208,7 +1209,18 @@ def parse_config(config_file: Path, platform, output_format, console):
 
     # Merge the PEP621 configuration (if it exists)
     try:
-        merge_pep621_config(global_config, pyproject["project"])
+        project_config = pyproject["project"]
+        if dynamic := project_config.get("dynamic"):
+            # Project defines dynamic PEP621 values, run its build backend to resolve them
+            metadata = project_wheel_metadata(base_path, isolated=True)
+            for key in dynamic:
+                # overwrite the dynamic values with their resolved counterparts
+                # TODO: key.title() is a poor substitute for a real translation
+                project_config[key] = metadata[key.title()]
+
+            project_config["dynamic"] = []
+
+        merge_pep621_config(global_config, project_config)
     except KeyError:
         pass
 ```

**Comment by freakboy3742** [comment]
The general approach definitely seems worth exploring.

My concern is whether this change by itself is enough. This approach will only work if you *also* have a PEP 517 `build-system` configuration for your project. That means either:
1. Adding a setuptools (or equivalent) configuration to your project even though generating a wheel isn't necessarily a valid option for your project, or
2. Adding a valid PEP 517 interface to Briefcase.

(1) is the "simple" solution; but I'd argue (2) is the "right" approach. 

I guess we could accept a (1)-based solution on the basis that it solves the immediate problem, but it would make for some awkward documentation. My preference would be to do this "right" from the outset.

**Comment by mhsmith** [comment]
After discussion in person, I think we agreed that (1) is fine, as it gives the user the flexibility to use whatever dynamic metadata source they like as long as it can return its result through `project_wheel_metadata`.

**Comment by akaIDIOT** [comment]
That's great to hear, that would be my preference too (the project I'm wrestling with is mainly a wheelable thing, a binary build to install on windows is but a side effect). Is this something you'll work towards yourself, or would you like me to polish up the patch to a state that would be suitable to a pull request?

**Comment by mhsmith** [comment]
We don't have any immediate plans to work on this, so a pull request would  be welcome.

**Comment by freakboy3742** [comment]
Fixed by #2772.
